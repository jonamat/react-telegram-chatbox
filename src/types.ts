import { ElementType } from 'react';

export interface WidgetProps {
    /** Title of the widget */
    title?: string;
    /** Subtitle of the widget */
    subtitle?: string;
    /** The placeholder of the message input */
    senderPlaceHolder?: string;
    /** The profile image that will be set on the responses */
    profileAvatar?: string;
    /** The picture image that will be shown next to the chat title */
    titleAvatar?: string;
    /** Show or hide the close button in full screen mode */
    showCloseButton?: boolean;
    /** Allow the use of full screen in full desktop mode */
    fullScreenMode?: boolean;
    /** Autofocus or not the user input */
    autofocus?: boolean;
    /** Custom Launcher component to use instead of the default */
    launcher?: (handleToggle: any) => ElementType;
    /** Function to handle the user clicking a quick button, will receive the 'value' when clicked. */
    handleQuickButtonClicked?: any;
    /** Show time stamp on messages */
    showTimeStamp?: boolean;
    /** Chat container id for a11y */
    chatId?: string;
    /** Alt value for the launcher when closed */
    launcherOpenLabel?: string;
    /** Alt value for the launcher when open */
    launcherCloseLabel?: string;
    /** Send button alt for a11y purposes */
    sendButtonAlt?: string;
    /** Prop that triggers on input change */
    handleTextInputChange?: (event: any) => any;
    /** Prop that triggers when a message is submitted, used for custom validation */
    handleSubmit?: (event: any) => any;
}

export interface ChatboxProps {
    /** Token generated by TgSocketBot */
    token: string;
    /** Send a message on chatbox creation */
    welcomeMessage?: string;
    /** Delay for the welcomeMessage */
    welcomeMessageDelay?: number;
    /** Message to send if connection fall or cannot be established */
    errorMessage?: string;
    /** Props for react-chat-widget */
    widgetProps?: WidgetProps;
}
