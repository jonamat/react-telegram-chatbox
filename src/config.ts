export const TBS_HOST = 'http://telegram-socket-bot.tk';
export const TBS_PATH = '/socket';
export const DEF_ERROR_MESSAGE = 'Cannot connect to server';
