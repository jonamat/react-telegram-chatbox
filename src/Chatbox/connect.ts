import { toggleInputDisabled, addResponseMessage } from 'react-chat-widget';
import SocketIOClient from 'socket.io-client';

import { TBS_HOST, TBS_PATH } from '../config';

const connect = (token) =>
    new Promise<SocketIOClient.Socket>((res, rej) => {
        // Disallow other messages until connection is established
        toggleInputDisabled();

        const socket = SocketIOClient.connect(TBS_HOST, {
            path: TBS_PATH,
            // @ts-expect-error socketIO missing declaration
            auth: { token },
        });

        socket.on('connect', () => {
            socket.on('message', (message: string) => {
                addResponseMessage(message);
            });

            // Re-enable input
            toggleInputDisabled();
            res(socket);
        });

        socket.on('connect_error', rej);
    });

export default connect;
