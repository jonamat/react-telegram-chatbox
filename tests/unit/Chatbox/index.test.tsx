/* ----------------------------- Unit test suite ---------------------------- *
 *
 * Target: Chatbox component
 * Tests:
 * - Props validation
 * - Functionality
 * - Matching render snapshot
 *
 * -------------------------------------------------------------------------- */

import React from 'react';
import { render, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import userEvent from '@testing-library/user-event';
import { create } from 'react-test-renderer';
import { mocked } from 'ts-jest/utils';
import SocketIOClient from 'socket.io-client';

import connect from '@App/Chatbox/connect';
import { ChatboxProps } from '@App/types';
import { DEF_ERROR_MESSAGE } from '@App/config';

/* -------------------------------------------------------------------------- */
/*                                    Mocks                                   */
/* -------------------------------------------------------------------------- */

const mockedConnect = mocked(connect);
jest.mock('@App/Chatbox/connect', () => jest.fn(() => Promise.resolve({ send: () => undefined })));

const mockedSocketConnect = mocked(SocketIOClient.connect);
jest.mock('socket.io-client', () => ({
    connect: jest.fn(),
}));

const token = 'fakeToken';

/* -------------------------------------------------------------------------- */
/*                              Bridge operations                             */
/* -------------------------------------------------------------------------- */

beforeEach(() => {
    jest.clearAllMocks();
});

/* -------------------------------------------------------------------------- */
/*                               Render wrapper                               */
/* -------------------------------------------------------------------------- */

/**
 * Render the component with the custom configuration
 * @param props Component props
 */
const renderComponent = (props?: Partial<ChatboxProps>) => {
    let Chatbox;

    // Isolation is required for react-chat-widget
    jest.isolateModules(() => {
        Chatbox = require('@App/index').Chatbox;
    });

    return render(<Chatbox token={token} {...props} />);
};

/* -------------------------------------------------------------------------- */
/*                                    Tests                                   */
/* -------------------------------------------------------------------------- */

describe('Chatbox component', () => {
    it('renders properly', () => {
        expect(() => {
            renderComponent();
        }).not.toThrowError();
    });
    it('does not open connection on component creation', () => {
        renderComponent();

        expect(mockedSocketConnect).not.toBeCalled();
    });
    it('opens connection on first message', async () => {
        const { getByAltText, getByPlaceholderText } = renderComponent({
            widgetProps: { autofocus: true },
        });
        userEvent.click(getByAltText('Close chat'));

        /*
            JSDOM submit event does not provide submitters
            --> https://github.com/jsdom/jsdom/issues/3117

            userEvent.type(getByPlaceholderText('Type a message...'), 'first message!');
            userEvent.click(getByAltText('Send'));

            Force submit with a custom event
        */

        const workaround = getByPlaceholderText('Type a message...') as HTMLInputElement & { message: any };
        workaround.message = {
            value: 'first message!',
        };
        workaround.dispatchEvent(new window.Event('submit', { bubbles: true }));

        expect(mockedConnect).toBeCalledTimes(1);
    });
    it('shows welcome message if defined', () => {
        const { queryByText, getByAltText } = renderComponent({
            welcomeMessage: 'hi!',
        });

        // Open widget
        userEvent.click(getByAltText('Close chat'));

        expect(queryByText('hi!')).toBeTruthy();
    });
    it('shows welcome message after delay if welcomeMessageDelay is defined', () => {
        jest.useFakeTimers();
        const { queryByText, getByAltText } = renderComponent({
            welcomeMessage: 'hi!',
            welcomeMessageDelay: 1000,
        });

        // Open widget
        userEvent.click(getByAltText('Close chat'));

        expect(queryByText('hi!')).toBeFalsy();

        jest.advanceTimersByTime(1000);

        expect(queryByText('hi!')).toBeTruthy();
    });
    it('uses errorMessage if defined', async () => {
        mockedConnect.mockImplementationOnce(() => Promise.reject('connection lost'));
        const { queryByText, getByAltText, getByPlaceholderText } = renderComponent({
            errorMessage: 'custom',
        });

        // Open widget
        userEvent.click(getByAltText('Close chat'));

        const workaround = getByPlaceholderText('Type a message...') as HTMLInputElement & { message: any };
        workaround.message = {
            value: 'first message!',
        };
        workaround.dispatchEvent(new window.Event('submit', { bubbles: true }));

        await waitFor(() => expect(queryByText('custom')).toBeTruthy());
    });
    it('uses default error if errorMessage is undefined', async () => {
        mockedConnect.mockImplementationOnce(() => Promise.reject('connection lost'));
        const { queryByText, getByAltText, getByPlaceholderText } = renderComponent();

        // Open widget
        userEvent.click(getByAltText('Close chat'));

        const workaround = getByPlaceholderText('Type a message...') as HTMLInputElement & { message: any };
        workaround.message = {
            value: 'first message!',
        };
        workaround.dispatchEvent(new window.Event('submit', { bubbles: true }));

        await waitFor(() => expect(queryByText(DEF_ERROR_MESSAGE)).toBeTruthy());
    });
    it('passes widgetProps to Widget component', () => {
        const { queryByText, getByAltText, getByPlaceholderText } = renderComponent({
            widgetProps: {
                title: 'custom title',
            },
        });

        // Open widget
        userEvent.click(getByAltText('Close chat'));

        expect(queryByText('custom title')).toBeTruthy();
    });
    it('validates props', () => {
        console.error = jest.fn();

        // @ts-expect-error validation test
        renderComponent({ errorMessage: 123 });
        // @ts-expect-error validation test
        renderComponent({ token: 123 });
        // @ts-expect-error validation test
        renderComponent({ welcomeMessage: 123 });
        // @ts-expect-error validation test
        renderComponent({ welcomeMessageDelay: '123' });
        // @ts-expect-error validation test
        renderComponent({ widgetProps: 123 });

        expect(console.error).toBeCalledTimes(5);
    });
    it('matches the snapshot', () => {
        const snapshot = create(require('@App/index').Chatbox).toJSON();
        expect(snapshot).toMatchSnapshot();
    });
});
