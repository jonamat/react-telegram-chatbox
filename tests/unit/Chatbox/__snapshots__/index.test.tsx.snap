// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chatbox component matches the snapshot 1`] = `null`;
